app/Services/RouteSuggester.php

```diff
@@ -0,0 +1,142 @@
+        $dijkstra = new Dijkstra($venuesGraph);
+        $path = $dijkstra->shortestPaths('start', 'end');
+
+        if (count($path) > 0) {
+            $ids = $path[0];
+            unset($ids[0]);
+            unset($ids[count($ids)]);
+
+            $venues = Venue::whereIn('id', $ids)->get();
+            return $venues;
+        } else return false;
+    }
+
+    private function getLocalVenuesByAttributes(array $stops, array $start, array $end, float $searchRange)
+    {
+        $lat = $this->sort($start[0], $end[0]);
+        $long = $this->sort($start[1], $end[1]);
+
+        $venues = [];
+        //for each stop -> get venue ids where attributes
+        foreach ($stops as $attributes) {
+            $matches =
+                Venue::where(
+                    function (Builder $query) use ($lat, $long, $searchRange, $attributes) {
+                        $query
+                            ->whereHas('address', function (Builder $query) use ($lat, $long, $searchRange) {
+                                $query
+                                    ->where('latitude', '>', $lat[0] - $searchRange)
+                                    ->where('latitude', '<', $lat[1] + $searchRange)
+                                    ->where('longitude', '>', $long[0] - $searchRange)
+                                    ->where('longitude', '<', $long[1] + $searchRange);
+                            })
+                            ->whereHas('attributes', function (Builder $query) use ($attributes) {
+                                $query
+                                    ->whereIn('name', $attributes);
+                            }, '>=', count($attributes));
+                    }
+                )
+                ->join('addresses', 'venues.id', '=', 'venue_id')
+                ->get();
+
+            array_push($venues, $matches);
+        }
+        return $venues;
+    }
+
+    private function createVenuesGraph(array $stops)
+    {
+        $distanceOffset = 0.02;
+        $numberOfStops = count($stops);
+        $graph = array();
+
+        $start_vertices = array();
+        foreach ($stops[0] as $venue) {
+            $distance =
+                $this->calculateDistance(
+                    $this->start,
+                    [$venue->latitude, $venue->longitude]
+                );
+            $id = $venue->venue_id;
+            $start_vertices[$id] = $distance - $distanceOffset;
+        }
+        $graph['start'] = $start_vertices;
+
+        //last venues adj to end
+        foreach ($stops[$numberOfStops - 1] as $venue) {
+            $distance =
+                $this->calculateDistance(
+                    [$venue->latitude, $venue->longitude],
+                    $this->end
+                );
+            $id = $venue->venue_id;
+            $graph[$id] = ['end' => $distance - $distanceOffset];
+        }
+
+        //else current stop vertices adj to next stop vertices
```

Comment:
a lot going on in this function can we break out the logic a bit more for clarity?

Reply:
Ya agree